### PR TITLE
Fix crash when projection fails

### PR DESCRIPTION
**Pull request by xlqian**
asgard crashed when one of coord projection fails...

We should still return 'unreached' when projection fails.... 

But Valhalla's SourceToTarget takes only valid projections... 

What I've done, take only valid projection for Valhalla's SourceToTarget and construct the whole response by inserting failed projections...

### CODE REVIEW

File: asgard/handler.cpp [hunk 1]
```diff
@@ -30,21 +30,26 @@
...
+#include <utility>
 
 using namespace valhalla;
 
 namespace asgard {
 
+using ValhallaLocations = google::protobuf::RepeatedPtrField<valhalla::Location>;
+using ProjectedLocations = std::unordered_map<std::string, valhalla::baldr::PathLocation>;
+
+constexpr size_t MAX_MASK_SIZE = 10000;
```
Comment (AurelienLP) [inline asgard/handler.cpp]: Could you put a comment here on why 10000 please?

File: asgard/handler.cpp [hunk 2]
```diff
@@ -87,17 +92,25 @@ std::vector<std::string> get_locations_from_matrix_request(const LocationContext
...
-using ProjectedLocations = std::unordered_map<std::string, valhalla::baldr::PathLocation>;
-ValhallaLocations make_valhalla_locations_from_projected_locations(const std::vector<std::string>& navitia_locations,
-                                                                   const ProjectedLocations& projected_locations,
-                                                                   valhalla::baldr::GraphReader& graph) {
+std::pair<ValhallaLocations, ProjectionFailedMask>
+make_valhalla_locations_from_projected_locations(const std::vector<std::string>& navitia_locations,
+                                                 const ProjectedLocations& projected_locations,
+                                                 valhalla::baldr::GraphReader& graph) {
     ValhallaLocations valhalla_locations;
+    ProjectionFailedMask projection_mask;
```
Comment (AurelienLP) [inline asgard/handler.cpp]: Could you let a comment here explaning with words how you use `projection_mask` please ?

File: asgard/handler.cpp [hunk 3]
```diff
@@ -137,38 +150,75 @@ pbnavitia::Response Handler::handle_matrix(const pbnavitia::Request& request) {
...
+    LOG_INFO("Projection Done");
 
     LOG_INFO("Computing matrix...");
-    const auto res = matrix.SourceToTarget(valhalla_location_sources,
-                                           valhalla_location_targets,
-                                           graph,
-                                           mode_costing.get_costing(),
-                                           util::convert_navitia_to_valhalla_mode(mode),
-                                           get_distance(mode, request.sn_routing_matrix().max_duration()));
+    auto res = matrix.SourceToTarget(valhalla_location_sources,
```
Comment (AurelienLP) [inline asgard/handler.cpp]: `const`

File: asgard/handler.cpp [hunk 4]
```diff
@@ -137,38 +150,75 @@ pbnavitia::Response Handler::handle_matrix(const pbnavitia::Request& request) {
...
+                                     valhalla_location_targets,
+                                     graph,
+                                     mode_costing.get_costing(),
+                                     util::convert_navitia_to_valhalla_mode(mode),
+                                     get_distance(mode, request.sn_routing_matrix().max_duration()));
     LOG_INFO("Computing matrix done.");
 
     pbnavitia::Response response;
     int nb_unreached = 0;
+
```
Comment (AurelienLP) [inline asgard/handler.cpp]: Necessary ?

File: asgard/handler.cpp [hunk 5]
```diff
@@ -137,38 +150,75 @@ pbnavitia::Response Handler::handle_matrix(const pbnavitia::Request& request) {
...
+    const auto& failed_projection_mask = (navitia_sources.size() == 1) ? projection_mask_targets : projection_mask_sources;
+
+    size_t elt_idx = -1;
+    size_t resp_row_size = navitia_sources.size() == 1 ? navitia_targets.size() : navitia_sources.size();
+
+    assert(resp_row_size == failed_projection_mask.count() + res.size());
+
+    auto elt_it = res.cbegin();
+
+    while (++elt_idx < resp_row_size) {
```
Comment (AurelienLP) [inline asgard/handler.cpp]: `elt_idx` and `elt_it` can be confusing here... You could maybe find other names?

Comment (xlqian) [reply asgard/handler.cpp]: done

**Pull request merged**